### ISSUE TITLE
about unique improper torsion

### ISSUE THREAD

**Issue by swillow** [pandegroup/openmm#2315]
After I have looked at the code of 'forcefield.py' in the directory '../simtk/openmm/app', I have one question:
--- (from forcefield.py)-----
(line number) [code]
1224        \# Make a list of all unique improper torsions
1225
1226        for atom in range(len(bondedToAtom)):
1227            bondedTo = bondedToAtom[atom]
1228            if len(bondedTo) > 2:
1229                for subset in itertools.combinations(bondedTo, 3):
1230                    data.impropers.append((atom, subset[0], subset[1], subset[2]))

A constraint for the unique improper torsion might be len(bondedTo) == 3 instead of len(bondedTo) > 2 (see line 1228).

**Comment by peastman** [comment]
Why?  Is there a case where it isn't working correctly?

**Comment by swillow** [comment]
Dear Pandegroup/Openmm

Since simulatePdb.py (see example) failed (rarely but sometimes) to
reproduce the potential energy obtained from simulateAmber.py,
I have tried to figure out why.
I attached one example files showing the disagreement between
simulatePdb.py and simulateAmber.py.

Thank you for reading my e-mail and responding of my question.

Best regards,

Soohaeng

**Comment by peastman** [comment]
Github doesn't include attachments.  Can you post a link to it?

**Comment by swillow** [comment]
You might get the attached file from the following link:
https://drive.google.com/open?id=1EKTlUMKBPgfXEUagquFqt-X-QZMP6yOF

**Comment by swillow** [comment]
Greeting!

I tried to figure out why both amber.prmtop and openmm (pdb + forcefields) give different potential energies.
Firstly, I could see that the bonds generated by `topology.createStandardBonds()' differs from the bonds in the receptor.prmtop. 
For example, the receptor.prmtop has 3760 bonds, while the openmm (pdb + forcefields) generates 3680 bonds.
I have checked where was the missing bonds in the openmm (pdb +forcefields).
Based on the attached files, in which --28 Val--29 Val -- 30 Val--, all the bonds for 29 Val are missing in the openmm (pdb + forcefields). In the same way, (--44 Gly -- 45 Gly -- 46 Gly--), all the bonds for 45 Gly are missing.

Please check whether my observation is correct.

**Comment by swillow** [comment]
Dear Pandegroup/Openmm

I made a code to generate a prmtop file from a pdb file with the purpose of debugging where the error occurs.

https://github.com/swillow/pdb2amber

My code (_forcefield and pdb2amber) might help you.

Best regards,

Soohaeng

**Comment by peastman** [comment]
I see you closed the issue.  Does that mean you figured out the problem?

**Comment by swillow** [comment]
Dear Peter

I tried to find where the error occurs in  `forcefield.py', but I could not.
Instead of the modification of 'forcefield.py' within the openmm,
I separated this forcefield.py from the openmm and modified it for the
purpose to check whether the amber prmtop file can be generated from a pdb
file. (Here, I have to reproduce the lists of bonds, angles, torsions, and
impropers). Eventually, I could generate the amber prmtop file using a
modified forcefield.py.

Besides that, once I have my own forcefield for the iron-sulfur clusters,
that can be obtained from QM calculations,
I will generate the amber prmtop file of the iron-sulfur clusters using my
code.
After that, I will use the openmm with the amber inpcrd and prmtop files.

Best regards,

Soohaeng
